[stdlib] speed up `reversed()` of `Sequence`

valeriyvan opened the pull request:
<!-- What's in this pull request? -->
Speeds up reversing of sequences which provide reasonable `underestimatedCount`.
For sequence with count 1000  and underestimatedCount 990 speed up is 7 times.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

valeriyvan committed:
Speeds up reversing of sequences which provide reasonable

valeriyvan committed:
Renames closure parameter

CodaFi commented:
@swift-ci benchmark

airspeedswift commented on stdlib/public/core/SequenceAlgorithms.swift:
```diff
@@ -707,15 +707,29 @@ extension Sequence {
+        initedCount = underestimatedCount
+      }
+      while let next = iterator.next() {
+        result.insert(next, at: 0)
```
This will turn sequences that provide a non-zero underestimate into a quadratic operation for the undercounted elements. This might be rare, but I don't think we can do that. If it is rare, a reasonable compromise may be to revert to building the array in full at this point and then reversing it.

valeriyvan replied:
You are right. I will fix this.

airspeedswift commented on stdlib/public/core/SequenceAlgorithms.swift:
```diff
@@ -707,15 +707,29 @@ extension Sequence {
+    let underestimatedCount = self.underestimatedCount
+    if underestimatedCount == 0 {
+      var result = Array(self)
+      let count = result.count
```
I realize the old code did it too, but this could be simplified to just `result.reverse()` from this point.

airspeedswift commented on stdlib/public/core/SequenceAlgorithms.swift:
```diff
@@ -707,15 +707,29 @@ extension Sequence {
+      return result
+    } else {
+      var iterator = makeIterator()
+      var result = [Element](unsafeUninitializedCapacity: underestimatedCount) { buf, initedCount in
```
preferred style in the std lib is to avoid the `[Element]` sugar when calling the initializer. Prefer `var result: [Element] = .init(` or `= Array(`

valeriyvan replied:
Fixed

airspeedswift reviewed (changes requested):
I think we need to fix the potential quadratic performance issue here before we can ship it.

CodaFi commented:
These benchmark results look like noise - except the code size regression. I can try again, but were you running a particular set when you came up with that 7x figure?

airspeedswift commented:
I'm guessing the speedup was measured by hand? We may want to add a new benchmark that tests this performance.

The code size increases are fine I think, they're reflecting the additional logic in microbenchmarks of reverse.

xwu commented on stdlib/public/core/SequenceAlgorithms.swift:
```diff
@@ -707,15 +707,29 @@ extension Sequence {
+      var result = Array(self)
+      let count = result.count
+      for i in 0..<count/2 {
+        result.swapAt(i, count - ((i + 1) as Int))      }
```
Nit: there is something odd about the formatting of this line.

valeriyvan replied:
Fixed.

xwu commented on stdlib/public/core/SequenceAlgorithms.swift:
```diff
@@ -707,15 +707,29 @@ extension Sequence {
+      return result
+    } else {
+      var iterator = makeIterator()
+      var result = [Element](unsafeUninitializedCapacity: underestimatedCount) { buf, initedCount in
```
Nit: I think we're still respecting the 80-column formatting rule where possible. Also, "inited" is not an idiomatic English word, nor is it easily pronounceable in English; prefer "initialized."

valeriyvan replied:
Fixed.

xwu commented on stdlib/public/core/SequenceAlgorithms.swift:
```diff
@@ -707,15 +707,29 @@ extension Sequence {
+      var result = [Element](unsafeUninitializedCapacity: underestimatedCount) { buf, initedCount in
+        for i in 0..<underestimatedCount {
+          guard let next = iterator.next() else {
+              preconditionFailure("underestimatedCount greater than count")
```
The internal standard library equivalent is `_preconditionFailure`. But here, I think the more appropriate choice would be `_internalInvariantFailure` no?

valeriyvan replied:
Fixed for using `_internalInvariantFailure`.

valeriyvan committed:
Formatting, indentation.
Parameter initedCount renamed into initializedCount.
_internalInvariantFailure is used instead of preconditionFailure.

valeriyvan commented:
@airspeedswift Yes, I have tested manually in playground. I will test proper tests during day or two.

airspeedswift commented:
>  Yes, I have tested manually in playground

Playgrounds aren't a suitable technology for testing performance like this, as the code will not be optimized. I'd advice writing small main.swift-based benchmarks to test this kind of performance. I'd also recommend posting the code used and some benchmarks of different kinds of collection (i.e. a custom collection that underreports its count). You may want to check out [Attabench](https://github.com/attaswift/Attabench) which is a good way of testing Swift code at different scales of input.

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

valeriyvan committed:
Simplify codepath with underestimatedCount == 0

valeriyvan committed:
Address quadratic compexity reversing undercounted ellements

valeriyvan committed:
Make result let constant

valeriyvan committed:
Add comment
